### PR TITLE
CC-350, CC-1097: Added support for delete and null handling

### PR DESCRIPTION
**Pull request by rhauch**
Added a configuration to control how records with null values (e.g., tombstone messages) are handled, with three options: `IGNORE`, `DELETE`, and `FAIL`.

The `FAIL` option is identical to the previous behavior. The `DELETE` option generates a corresponding delete operation in the Elasticsearch system only when the record has an explicit key; if `key.ignore` is true, the keys will be generated and delete does not make sense. And finally the `IGNORE` option tells the connector to quietly skip over any records that have a null value.

Although the previous behavior is equivalent to `FAIL`, that cause the connector for fail and stop whenever it came across a null value. A more sensible default is `IGNORE`.

NOTE: This is an alternative PR with nearly the same changes as #163 but with fewer commits and without the other formatting/style changes.

### CODE REVIEW

File: src/main/java/io/confluent/connect/elasticsearch/DataConverter.java [hunk 1]
```diff
@@ -242,4 +303,51 @@ Object preProcessValue(Object value, Schema schema, Schema newSchema) {
         return value;
     }
   }
+  public enum BehaviorOnNullValues {
+    IGNORE,
+    DELETE,
+    FAIL;
+
+    public static final BehaviorOnNullValues DEFAULT = IGNORE;
```
Comment (C0urante) [inline src/main/java/io/confluent/connect/elasticsearch/DataConverter.java]: If we're aiming for backwards compatibility ("The FAIL option is identical to the previous behavior, so this option is the default to ensure backward compatibility for existing connector deployments that are upgrading") this should be changed to `DEFAULT = FAIL;`.

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java [hunk 2]
```diff
@@ -220,7 +238,19 @@ public void write(Collection<SinkRecord> records) {
       final IndexableRecord indexableRecord = converter.convertRecord(sinkRecord, index, type,
                                                                       ignoreKey, ignoreSchema);
 
-      bulkProcessor.add(indexableRecord, flushTimeoutMs);
+      // In the event that the sink record's value was null and the data converter has been told to
+      // ignore null values, the returned record will be null.
```
Comment (C0urante) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]: Had a `TODO` here just in case it turned out it might be advantageous for the `ElasticsearchWriter` to intercept null-valued records before this point when `behavior.on.null.values` is set to `IGNORE`, in order to avoid possible performance penalties and/or other issues arising from the `if (!ignoreSchema && !existingMappings.contains(index))` block. Should that be put back in or was it removed intentionally?

File: src/main/java/io/confluent/connect/elasticsearch/DataConverter.java (hunk 1 above)
Comment (kkonstantine) [reply src/main/java/io/confluent/connect/elasticsearch/DataConverter.java]: I think it's safe to enable a sane config as the default (here `IGNORE`). Backwards compatibility, as I see it here, means that if people have built software around the assumption that the connector fails, they can still enable this behavior. 

But we should have the connector, starting with this version, not failing when it encounters records with non-null key/null value.

File: src/test/java/io/confluent/connect/elasticsearch/MappingTest.java [hunk 3]
```diff
@@ -109,7 +111,7 @@ private void verifyMapping(Schema schema, JsonObject mapping) throws Exception {
       }
     }
 
-    DataConverter converter = new DataConverter(true);
+    DataConverter converter = new DataConverter(true, BehaviorOnNullValues.FAIL);
```
Comment (kkonstantine) [inline src/test/java/io/confluent/connect/elasticsearch/MappingTest.java]: Do we want the current default here (`IGNORE`)?

Comment (C0urante) [reply src/test/java/io/confluent/connect/elasticsearch/MappingTest.java]: No, this should be `FAIL` since there aren't any apparent cases for passing a null-valued record to the converter in this test.

File: src/main/java/io/confluent/connect/elasticsearch/DataConverter.java (hunk 1 above)
Comment (rhauch) [reply src/main/java/io/confluent/connect/elasticsearch/DataConverter.java]: I'm fine with changing the documentation, since the previous behavior is not really a useful default. @C0urante, do you want to make those changes and push them to this branch?

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java (hunk 2 above)
Comment (rhauch) [reply src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]: I removed it intentionally. We can change the code now, but I don't think the speculative TODO was useful.

**Comment by rhauch** [comment]
Correctly based on `3.3.x` rather than `master`.

File: src/main/java/io/confluent/connect/elasticsearch/DataConverter.java (hunk 1 above)
Comment (C0urante) [reply src/main/java/io/confluent/connect/elasticsearch/DataConverter.java]: Which documentation changes are you referring to? `configuration_options.rst` already lists `ignore` as the default, not sure where else would need to be changed that hasn't already been.

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java (hunk 2 above)
Comment (C0urante) [reply src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]: Alright, fair enough. Guess it can't hurt to change it now--as conceptually pleasing as it is to [dr]elegate (😜) handling of null-valued records to the `DataConverter` class I see some calls to the `JestClient` inside the for-loop above that would probably be beneficial to avoid.

File: src/main/java/io/confluent/connect/elasticsearch/DataConverter.java (hunk 1 above)
Comment (rhauch) [reply src/main/java/io/confluent/connect/elasticsearch/DataConverter.java]: Sorry, that comment was in the commit. Let me amend the commit message.

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java (hunk 2 above)
Comment (rhauch) [reply src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]: Let me know if you have problems pushing to this branch.

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java [hunk 4]
```diff
@@ -218,6 +220,11 @@ public ElasticsearchWriter build() {
 
   public void write(Collection<SinkRecord> records) {
     for (SinkRecord sinkRecord : records) {
+      // Preemptively skip records with null values if they're going to be ignored anyways
+      if (sinkRecord.value() == null && behaviorOnNullValues.equals(BehaviorOnNullValues.IGNORE)) {
```
Comment (rhauch) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]: How about `behaviorOnNullValues == BehaviorOnNullValues.IGNORE`? Enums are more like primitives than like objects.

**Comment by rhauch** [comment]
LGTM, but I cannot approve the PR since I created it. @C0urante  and @kkonstantine, please approve or comment.

File: src/test/java/io/confluent/connect/elasticsearch/MappingTest.java (hunk 3 above)
Comment (kkonstantine) [reply src/test/java/io/confluent/connect/elasticsearch/MappingTest.java]: I agree. `null` value is not tested here. Still, since we need to give an argument there, we should rather give the default, than one that would be used less frequently

Comment (rhauch) [reply src/test/java/io/confluent/connect/elasticsearch/MappingTest.java]: The `DataConverterTest` verifies that `DataConverter` behaves correctly. This test is merely using a `DataConverter` to check the mapping, and it seems like we'd want the test to fail if there are any unexpected problems. How would using a different `BehaviorOnNullValue` affect this test?

Comment (kkonstantine) [reply src/test/java/io/confluent/connect/elasticsearch/MappingTest.java]: I don't think it would affect the test, since this test doesn't seem to check the case of record with a `null` value. My point is that we should not depend on testing something "by accident". If we need a test case for `null` we should add one.
